### PR TITLE
 Adds notifications to cypher when using a suboptimal index seek

### PR DESCRIPTION
**Pull request by sherfert**
changelog:
When writing WHERE n.str CONTAINS 'x'
          or WHERE n.str ENDS WITH 'x'
Cypher will issue a warning, if n.str is backed by a native String index
which is suboptimal for these index scans.

### CODE REVIEW

**Comment by burqen** [comment]
@sherfert Awesome! Is this enough to display warning in browser or is there more to be done?

**Comment by sherfert** [comment]
Good question @burqen . I will verify if this suffices.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/LogicalPlanNotifications.scala [hunk 1]
```diff
@@ -17,19 +17,25 @@
...
+  def apply(plan: LogicalPlan): TraversableOnce[InternalNotification]
+}
+
 object LogicalPlanNotifications {
   def checkForNotifications(logicalPlan: LogicalPlan,
                             planContext: PlanContext,
                             config: CypherCompilerConfiguration): Seq[InternalNotification] = {
-    val notificationCheckers = Seq(checkForEagerLoadCsv,
-      CheckForLoadCsvAndMatchOnLargeLabel(planContext, config.nonIndexedLabelWarningThreshold))
+    val notificationCheckers: Seq[NotificationChecker] = Seq(
```
Comment (Lojjs) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/LogicalPlanNotifications.scala]: No need to write out the type annotation here

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/LogicalPlanNotifications.scala [hunk 2]
```diff
@@ -17,19 +17,25 @@
...
-package org.neo4j.cypher.internal.compatibility.v3_4
+package org.neo4j.cypher.internal.compatibility.v3_4.notifications
 
 import org.neo4j.cypher.internal.compiler.v3_4.CypherCompilerConfiguration
 import org.neo4j.cypher.internal.frontend.v3_4.notification.InternalNotification
 import org.neo4j.cypher.internal.planner.v3_4.spi.PlanContext
 import org.neo4j.cypher.internal.v3_4.logical.plans.LogicalPlan
 
+trait NotificationChecker {
+  def apply(plan: LogicalPlan): TraversableOnce[InternalNotification]
```
Comment (Lojjs) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/LogicalPlanNotifications.scala:28]: I'm not familiar with `TraversableOnce`. What was the benefit of using this instead of `Option[NotificationChecker]`?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForEagerLoadCsv.scala [hunk 3]
```diff
@@ -17,14 +17,14 @@
...
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.cypher.internal.compatibility.v3_4
+package org.neo4j.cypher.internal.compatibility.v3_4.notifications
 
 import org.neo4j.cypher.internal.frontend.v3_4.notification.{EagerLoadCsvNotification, InternalNotification}
 import org.neo4j.cypher.internal.v3_4.logical.plans.{Eager, LoadCSV, LogicalPlan}
 
-object checkForEagerLoadCsv extends (LogicalPlan => Option[InternalNotification]) {
+object checkForEagerLoadCsv extends NotificationChecker {
```
Comment (Lojjs) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForEagerLoadCsv.scala:25]: I know you haven't introduced this, but shouldn't Scala objects start with a capital letter? i.e. `CheckForEagerLoadCSV`

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForLoadCsvAndMatchOnLargeLabel.scala [hunk 4]
```diff
@@ -17,20 +17,20 @@
...
+package org.neo4j.cypher.internal.compatibility.v3_4.notifications
 
 import org.neo4j.cypher.internal.frontend.v3_4.notification.{InternalNotification, LargeLabelWithLoadCsvNotification}
 import org.neo4j.cypher.internal.planner.v3_4.spi.PlanContext
-import org.neo4j.cypher.internal.util.v3_4.{Cardinality, LabelId, NameId}
+import org.neo4j.cypher.internal.util.v3_4.{Cardinality, LabelId}
 import org.neo4j.cypher.internal.v3_4.logical.plans.{LoadCSV, LogicalPlan, NodeByLabelScan}
 
-case class CheckForLoadCsvAndMatchOnLargeLabel(planContext: PlanContext,
+case class checkForLoadCsvAndMatchOnLargeLabel(planContext: PlanContext,
```
Comment (Lojjs) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForLoadCsvAndMatchOnLargeLabel.scala:27]: Why did you change the name from `CheckForLoadCsvAndMatchOnLargeLabel` to `checkForLoadCsvAndMatchOnLargeLabel`? As far as I know, Scala case classes should start with a capital letter, just like normal classes.

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/notification/InternalNotification.scala [hunk 5]
```diff
@@ -70,3 +70,5 @@ case class DeprecatedRelTypeSeparatorNotification(position: InputPosition) exten
 case object DeprecatedPlannerNotification extends InternalNotification
 
 case class ExperimentalFeatureNotification(msg: String) extends InternalNotification
+
+case class SuboptimalIndexForWildcardQueryNotification(label: String, propertyKeys: Seq[String]) extends InternalNotification
```
Comment (Lojjs) [inline community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/notification/InternalNotification.scala]: It looks like you inserted some weird line break or similar here (look at the red symbol in the diff).

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForIndexLimitation.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ */
+package org.neo4j.cypher.internal.compatibility.v3_4.notifications
+
+import org.neo4j.cypher.internal.frontend.v3_4.notification.{InternalNotification, SuboptimalIndexForWildcardQueryNotification}
+import org.neo4j.cypher.internal.planner.v3_4.spi.{IndexLimitation, PlanContext, SlowContains}
+import org.neo4j.cypher.internal.util.v3_4.{Cardinality, LabelId}
+import org.neo4j.cypher.internal.v3_4.expressions.{LabelToken, PropertyKeyToken}
+import org.neo4j.cypher.internal.v3_4.logical.plans._
+
+case class checkForIndexLimitation(planContext: PlanContext) extends NotificationChecker {
```
Comment (Lojjs) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForIndexLimitation.scala:28]: Correct me if I'm wrong, but as far as I know, Scala case classes should begin with a capital letter, just like normal classes.

File: community/cypher/planner-spi-3.4/src/main/scala/org/neo4j/cypher/internal/planner/v3_4/spi/IndexDescriptor.scala [hunk 7]
```diff
@@ -21,19 +21,23 @@ package org.neo4j.cypher.internal.planner.v3_4.spi
 
 import org.neo4j.cypher.internal.util.v3_4.{LabelId, PropertyKeyId}
 
+sealed trait IndexLimitation
+case object SlowContains extends IndexLimitation
+
 object IndexDescriptor {
-  def apply(label: Int, property: Int): IndexDescriptor = IndexDescriptor(LabelId(label), Seq(PropertyKeyId(property)))
+  def apply(label: Int, property: Int): IndexDescriptor = IndexDescriptor(LabelId(label), Seq(PropertyKeyId(property)), Set.empty[IndexLimitation])
```
Comment (Lojjs) [inline community/cypher/planner-spi-3.4/src/main/scala/org/neo4j/cypher/internal/planner/v3_4/spi/IndexDescriptor.scala]: You don't have to change the 3 old apply methods to set `limitations` explicitly since `Set.empty[IndexLimitation]` already is the default value.

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NotificationAcceptanceTest.scala [hunk 8]
```diff
@@ -653,6 +653,72 @@ class NotificationAcceptanceTest extends ExecutionEngineFunSuite with CypherComp
...
+    val result = innerExecuteDeprecated(query, Map.empty)
+    result.notifications should contain(SUBOPTIMAL_INDEX_FOR_WILDCARD_QUERY.notification(graphdb.InputPosition.empty, suboptimalIndex("Person", "name")))
+  }
+
+  test("should warn when using ends with on a unique index with SLOW_CONTAINS limitation") {
+    graph.createConstraint("Person", "name")
+    val query = "EXPLAIN MATCH (a:Person) WHERE a.name ENDS WITH 'son' RETURN a"
+    val result = innerExecuteDeprecated(query, Map.empty)
+    result.notifications should contain(SUBOPTIMAL_INDEX_FOR_WILDCARD_QUERY.notification(graphdb.InputPosition.empty, suboptimalIndex("Person", "name")))
+  }
```
Comment (Lojjs) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NotificationAcceptanceTest.scala:683]: Overall good testing, but I'm thinking about if should write a test to check that the warning is NOT thrown when using `STARTS WITH`? Either in this file or in `CheckForIndexLimitationTest`. Do you think such a test would be of any value or is it is already covered in the kernel work?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForIndexLimitation.scala [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+          (acc ++ notifications, None)
+    }
+  }
+
+  def getLimitations(label: LabelToken, property: PropertyKeyToken): Set[IndexLimitation] = {
+    planContext.indexGet(label.name, Seq(property.name)).fold(Set.empty[IndexLimitation])(_.limitations) union
+    planContext.uniqueIndexGet(label.name, Seq(property.name)).fold(Set.empty[IndexLimitation])(_.limitations)
+  }
+
+  private def cardinality(labelName: String): Cardinality =
```
Comment (Lojjs) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForIndexLimitation.scala]: This method is never used, please remove it!

**Review by Lojjs — CHANGES_REQUESTED** [review]
The main things looks good, but I have a bunch of smaller comments. Please, go through and fix.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/LogicalPlanNotifications.scala (hunk 2 above)
Comment (sherfert) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/LogicalPlanNotifications.scala:28]: Option would not work because of the new `NotificationChecker` we added, that would have more than one Notification. `TraversableOnce` is just a very high level interface that allows us to do `flatMap`, which is all we want here. Then the implementations can choose themselves if `Seq` or `Iterator` or whatnot is the best choice.

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NotificationAcceptanceTest.scala (hunk 8 above)
Comment (sherfert) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NotificationAcceptanceTest.scala:683]: I think that is a good idea.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForEagerLoadCsv.scala (hunk 3 above)
Comment (sherfert) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForEagerLoadCsv.scala:25]: In general yes, but for objects that only define a public `apply` method and as such can be viewed as an object that encapsulates exactly one function, lower case is acceptable, too. We do this in a lot of places, e.g. `idSeekLeafPlanner`, `pickBestPlanUsingHintsAndCost`, etc.

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForLoadCsvAndMatchOnLargeLabel.scala (hunk 4 above)
Comment (sherfert) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForLoadCsvAndMatchOnLargeLabel.scala:27]: In this case for consistency with the other `NotificationChecker`s.

Comment (Lojjs) [reply community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/notifications/checkForLoadCsvAndMatchOnLargeLabel.scala:27]: Well, before there were 2 of them, one which started with a lowecase letter and one which started with uppercase. I like that you made it more consistent, but what you did was to change the uppercase to lowercase and create a new one, also lowercase. It would be exactly the same amount of change to make them all uppercase, which I think is more correct.

**Review by tinwelint — APPROVED** [review]
From what I can tell this looks to do the right thing
